### PR TITLE
[RELAY] [OP] [MXNet Frontend] Add sequence_mask

### PR DESCRIPTION
**Pull request by sxjscience**
1.  Support sequence_mask in relay (also topi).
2. Support MXNet frontend conversion of SequenceMask
3. Fix the support level of take to be 3. Same as the documentation: https://docs.tvm.ai/langref/relay_op.html#level-3-definitions

### CODE REVIEW

**Comment by icemelon9** [comment]
Could you also add this op in Relay and MxNet frontend converter?

**Comment by sxjscience** [comment]
OK, will do this today. How should I test the converter? I find I could add here https://github.com/dmlc/tvm/blob/master/python/tvm/relay/frontend/mxnet.py

**Comment by icemelon9** [comment]
The frontend test code is here https://github.com/dmlc/tvm/blob/master/tests/python/frontend/mxnet/test_forward.py

**Comment by junrushao1994** [comment]
Just a minor question: should it be put under transform.py?

**Comment by sxjscience** [comment]
I'm actually not sure which file I should put this op in. Since it's injective and somehow similar to `take` which is in transform.py, I also put it into the transform.py.

File: python/tvm/relay/frontend/mxnet.py [hunk 1]
```diff
@@ -709,6 +709,18 @@ def _mx_topk(inputs, attrs):
...
+def _mx_SequenceMask(inputs, attrs):
+    assert len(inputs) == 1 or len(inputs) == 2
+    new_attrs = {}
+    use_sequence_length = attrs.get_bool('use_sequence_length', False)
+    new_attrs['mask_value'] = attrs.get_float('value', 0.0)
+    new_attrs['axis'] = attrs.get_int('axis', 0)
+    if use_sequence_length:  #pylint: disable=no-else-return
+        return _op.sequence_mask(*inputs, **new_attrs)
+    else:
+        return _op.copy(inputs[0])
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py]:
```suggestion
        return inputs[0]
```

File: topi/python/topi/transform.py [hunk 2]
```diff
@@ -436,3 +437,44 @@ def shape(array, dtype="int32"):
...
+    `valid_length` gives the length of each sequence. `valid_length` should be
+     a 1D int array with positive ints and has dimension [batch_size,].
+
+    Parameters
+    ----------
+    data : tvm.Tensor
+        N-D with shape [MAX_LENGTH, batch_size, ...] or [batch_size, MAX_LENGTH, ...]
+         depending on the value of `axis`.
+
+    valid_length : tvm.Tensor or None
```
Comment (icemelon9) [inline topi/python/topi/transform.py]:
```suggestion
    valid_length : tvm.Tensor, optional
```

File: topi/python/topi/transform.py [hunk 3]
```diff
@@ -436,3 +437,44 @@ def shape(array, dtype="int32"):
...
+        1-D with shape [batch_size,]
+
+    mask_value : float, default 0
```
Comment (icemelon9) [inline topi/python/topi/transform.py]:
```suggestion
    mask_value : float, optional
```

File: topi/python/topi/transform.py [hunk 4]
```diff
@@ -436,3 +437,44 @@ def shape(array, dtype="int32"):
...
+        The masking value, default
```
Comment (icemelon9) [inline topi/python/topi/transform.py]:
```suggestion
        The masking value, default 0
```

File: topi/python/topi/transform.py [hunk 5]
```diff
@@ -436,3 +437,44 @@ def shape(array, dtype="int32"):
...
+
+    axis : int, default 0
```
Comment (icemelon9) [inline topi/python/topi/transform.py]:
```suggestion
    axis : int, optional
```

File: topi/python/topi/transform.py [hunk 6]
```diff
@@ -436,3 +437,44 @@ def shape(array, dtype="int32"):
...
+        axis of the length dimension, must be 0 or 1.
+
+    Returns
+    -------
+    output : tvm.Tensor
+        N-D with shape [MAX_LENGTH, batch_size, ...] or [batch_size, MAX_LENGTH, ...]
+         depending on the value of `axis`.
```
Comment (icemelon9) [inline topi/python/topi/transform.py]: fix indent

File: src/relay/op/tensor/transform.cc [hunk 7]
```diff
@@ -2218,5 +2218,103 @@ output shape will simply be (Y_0, ..., Y_{K-1}).
...
+TVM_REGISTER_NODE_TYPE(SequenceMaskAttrs);
+
+bool SequenceMaskRel(const Array<Type>& types,
+                     int num_inputs,
+                     const Attrs& attrs,
+                     const TypeReporter& reporter) {
+  // `types` contains: [data, valid_length, result]
+  CHECK_EQ(types.size(), 3);
+  const auto* data = types[0].as<TensorTypeNode>();
+  CHECK(data != nullptr);
```
Comment (icemelon9) [inline src/relay/op/tensor/transform.cc]:
```suggestion
  CHECK(data);
```

File: src/relay/op/tensor/transform.cc [hunk 8]
```diff
@@ -2218,5 +2218,103 @@ output shape will simply be (Y_0, ..., Y_{K-1}).
...
+  const auto* valid_length = types[1].as<TensorTypeNode>();
```
Comment (icemelon9) [inline src/relay/op/tensor/transform.cc]: Change this to `reporter->Assign(types[1], valid_type);` where `valid_type` should be a tensor type of [batch_size].

File: python/tvm/relay/frontend/mxnet.py [hunk 9]
```diff
@@ -709,6 +709,18 @@ def _mx_topk(inputs, attrs):
     return _op.topk(inputs[0], **new_attrs)
 
 
+def _mx_SequenceMask(inputs, attrs):
+    assert len(inputs) == 1 or len(inputs) == 2
+    new_attrs = {}
+    use_sequence_length = attrs.get_bool('use_sequence_length', False)
+    new_attrs['mask_value'] = attrs.get_float('value', 0.0)
+    new_attrs['axis'] = attrs.get_int('axis', 0)
+    if use_sequence_length:  #pylint: disable=no-else-return
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py]: Move pylint disable to top of the file

File: src/relay/op/tensor/transform.cc [hunk 10]
```diff
@@ -2218,5 +2218,107 @@ output shape will simply be (Y_0, ..., Y_{K-1}).
...
+                     const TypeReporter& reporter) {
+  // `types` contains: [data, valid_length, result]
+  CHECK_EQ(types.size(), 3);
+  const auto* data = types[0].as<TensorTypeNode>();
+  const auto* valid_length = types[1].as<TensorTypeNode>();
+  CHECK(data);
+  CHECK(valid_length);
+  const auto param = attrs.as<SequenceMaskAttrs>();
+  Array<IndexExpr> valid_length_shape;
+  valid_length_shape.push_back(data->shape[1 - param->axis]);
```
Comment (icemelon9) [inline src/relay/op/tensor/transform.cc]: Add check for `param->axis` before this line.

Comment (sxjscience) [reply src/relay/op/tensor/transform.cc]: Added

**Comment by icemelon9** [comment]
Thanks @sxjscience . This is now merged.

**Pull request merged**